assets/js/atomic/utils/index.js

```diff
@@ -1,64 +1,39 @@
-import {
-	ProductListLink,
-	ProductListTitle,
-	ProductListPrice,
-	ProductListButton,
-	ProductListImage,
-	ProductListRating,
-	ProductListSummary,
-	ProductListSaleBadge,
-} from '../components/product-list';
+/**
+ * External dependencies
+ */
+import { applyFilters } from '@wordpress/hooks';
 
 /**
- * List of mapped components.
+ * Internal dependencies
  */
-export const COMPONENT_MAP = {
-	ProductListLink,
+import {
 	ProductListTitle,
 	ProductListPrice,
 	ProductListButton,
 	ProductListImage,
 	ProductListRating,
 	ProductListSummary,
 	ProductListSaleBadge,
-};
+} from '../components/product-list';
 
 // @todo how to support these using mapping?
 //'core/paragraph',
 //'core/heading',
 /**
  * Map blocks names to component names.
  */
-export const BLOCK_MAP = [
-	{
-		component: 'ProductListPrice',
-		blockName: 'woocommerce/product-list-price',
-	},
-	{
-		component: 'ProductListImage',
-		blockName: 'woocommerce/product-list-image',
-	},
-	{
-		component: 'ProductListTitle',
-		blockName: 'woocommerce/product-list-title',
-	},
-	{
-		component: 'ProductListRating',
-		blockName: 'woocommerce/product-list-rating',
-	},
-	{
-		component: 'ProductListButton',
-		blockName: 'woocommerce/product-list-button',
-	},
+export const BLOCK_MAP = applyFilters(
+	'woocommerce_blocks_product_grid_children',
 	{
-		component: 'ProductListSummary',
-		blockName: 'woocommerce/product-list-summary',
-	},
-	{
-		component: 'ProductListSaleBadge',
-		blockName: 'woocommerce/product-list-sale-badge',
-	},
-];
+		'woocommerce/product-list-price': ProductListPrice,
+		'woocommerce/product-list-image': ProductListImage,
+		'woocommerce/product-list-title': ProductListTitle,
+		'woocommerce/product-list-rating': ProductListRating,
+		'woocommerce/product-list-button': ProductListButton,
+		'woocommerce/product-list-summary': ProductListSummary,
+		'woocommerce/product-list-sale-badge': ProductListSaleBadge,
+	}
+);
```

Comment:
Do we want to expose the possibility for removing blocks from this object?   Should we protect certain blocks from being removed?  Or is this intentional so they can be replaced with enhanced options.

Having a filter seems a bit fragile because:

- existing blocks can be removed (which can introduce block errors for already created blocks with the existing inner blocks)
- If blocks are registered and then removed via a plugin, existing blocks will have an error (granted this is something that exists in core as well... but still something to keep in mind).

Reply:
> If blocks are registered and then removed via a plugin, existing blocks will have an error (granted this is something that exists in core as well... but still something to keep in mind).

This was something in the issue we need to accommodate since there are valid reasons why a block would go away. Skipping rendering when this happens would be my pref, vs an error.

Reply:
> Having a filter seems a bit fragile

This is probably the case with most filters in WP right? Any developer *could* break something if mishandling the data.

We could have the registration function as suggested, and document that as the prefered way, but inject what gets registered via the filter so it's there (or hold off on the filter, but work on the assumption it will be there eventually?)

Reply:
> This is probably the case with most filters in WP right? Any developer could break something if mishandling the data.

It is.  It's also good reason to try and preserve both extensibility plus reduce the likelihood of breakage for better user experience.

> We could have the registration function as suggested, and document that as the prefered way, but inject what gets registered via the filter so it's there (or hold off on the filter, but work on the assumption it will be there eventually?)

I'm thinking that we could start with encapsulated local state for holding the registered inner blocks and avoid the use of a filter all together.

Reply:
> I'm thinking that we could start with encapsulated local state for holding the registered inner blocks and avoid the use of a filter all together.

What do you mean by "_encapsulated local state_" @nerrad? Do you have an implementation idea in mind or should I start working on my own?

I like the idea of creating a `registerInnerBlocks()` function instead of using hooks, but would be happy to know your thoughts/ideas if you have any before starting with it. :slightly_smiling_face:

Reply:
@nerrad I [updated](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/3e17001e8ba3220c3ed0f272a86714659cac1d40) this PR based on your proposal, could you take a look? I updated the example extension too, so the steps to test should be the same.